llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3529,6 +3529,38 @@ static Instruction *foldBitCeil(SelectInst &SI, IRBuilderBase &Builder) {
                                 Masked);
 }
 
+// This function tries to fold the following sequence
+//   %lt = icmp ult/slt i32 %x, %y
+//   %ne0 = icmp ne i32 %x, %y
+//   %ne = zext i1 %ne0 to iN
+//   %r = select i1 %lt, iN -1, iN %ne
+// into
+//   %r = call iN @llvm.ucmp/scmp(%x, %y)
+Instruction *InstCombinerImpl::foldSelectToCmp(SelectInst &SI) {
+  if (!isa<Constant>(SI.getTrueValue()) ||
```

Comment:
Think you could be moving the true/false values to variables and swapping based on the condition. Otherwise you won't be handling the inverse case: https://godbolt.org/z/rE7PeYMae

Reply:
This case is incorrect. It returns `-1` when `%y == %x`.

Reply:
Sorry its `uge` not `ugt`: https://alive2.llvm.org/ce/z/TmFypc